### DISCUSSION TITLE
dagster daemon/dagit machine running out of memory

### DISCUSSION THREAD

**Discussion by mmarionncern** [dagster-io/dagster#15216] [Q&A]
Dear experts,

I am using dagster-celery with dockers for my computing cluster, and it appears regularly that the node supporting the dagster daemon docker, the dagit docker and a user code docker that is not used for computing runs out of memory with plenty of python processed that never ends

Sorry for the quality of the screenshot :
![dagster-memory](https://github.com/dagster-io/dagster/assets/5114248/ff5b013d-e442-4cf3-a516-ae55b50b2da3)

Is there something that can be done to clean the memory without shutting down/up the dagster dockers (which is the only way I found so far to avoid the memory saturation)?

Thank you by advance for your help

Matthieu

**Comment by alangenfeld** [comment]
I'm surprised to see `--socket /tmp/...` in the code server cli args. That piece of information leads me to believe that you have the webserver and/or daemons targeting python directly which causes them to spin up user code servers as subprocesses. 

Though based on 

> a user code docker 

it would seem you are trying to manage the code server as a container like covered in this example https://docs.dagster.io/deployment/guides/docker#multi-container-docker-deployment

Does your `workspace.yaml` target grpc servers directly like shown in that link or does it refer to python targets?

**Comment by mmarionncern** [comment]
Thank for your answer Alex,

To go deeper in detail, we are indeed using dockers in the following way : 
- 1 "dagit" docker launched with :
`dagster instance migrate
dagit -h 0.0.0.0 -p 3000 -f hpp/repository.py` 

- 1 "dagster-daemon" docker : 
`dagster-daemon run -f hpp/repository.py`

- several "dagster-celery" dockers typically with the following launch : 
`dagster-celery worker start -n ${hostname} --loglevel=info --app=dagster_celery.app -y dagster-celery.yaml -q $queue -i hpp.repository -- --pool=prefork --concurrency $concurrency`

For the latest, all dockers are called "user_code" and we have the following workspace.yaml which is I think quite similar with the example you provided : 
```
load_from:
  # Each entry here corresponds to a service in the docker-compose file that exposes user code.
  - grpc_server:
      host: user_code
      port: 4000
      location_name: "user_code"
```

Could it be related to the fact that all our "dagster-celery worker containers" are named "user_code"?

**Reply by alangenfeld** [reply]
apologies for missing this before 

You are launching `dagit` and the `daemon` with `-f hpp/repository.py` . This instructs each of these processes to launch their own user code servers as subprocesses targeting that python file resulting in higher memory consumption due to duplication. The workspace.yaml file you mention does not appear to be used.

**Comment by rubimpassos** [comment]
I had the same issue, so I decided to investigate the code of dagster_celery. What I understood is that apparently, every time a job is sent to the queue, the daemon maintains an infinite loop process for job monitoring. Therefore, for each job running, the daemon consumes more CPU power and memory, making it completely non-scalable. Based on my analysis, it is impossible to use dagster_celery in production in a scenario where hundreds to thousands of jobs are executed simultaneously because the daemon does not support replicas

**Reply by skeller88** [reply]
Dagster team, is this accurate? This would be unfortunate. @rubimpassos I haven't looked at the daemon code, but I would assume the daemon monitoring involves polling the runs db and/or events db for job updates? That seems lightweight, and could scale to thousands of tasks especially if the daemon polls for metadata about multiple jobs at once. But is there something more heavyweight that the daemon is doing?

**Reply by alangenfeld** [reply]
I find this statement confusing, and expect that some terms are being used for things that lead to confusion.

> so I decided to investigate the code of dagster_celery.

I am not sure why we are talking about `dagster-celery` as its use has no effect on the `dagster-daemon`. `dagster-celery` provides only the `celery_executor` which executes Dagster `Jobs` by delegating each "step" (op/asset) out via celery. 

>  What I understood is that apparently, every time a job is sent to the queue, the daemon maintains an infinite loop process for job monitoring

So interpreting these words in how they are generally used in Dagster, I don't think this is quite right. A `DagsterRun` is queued by adding it in a queued state to the runs table. The `dagster-daemon` process polls the table for these queued runs and then launches them via the instance's `RunLauncher`. Then that launched process is responsible for orchestrating all of the steps (ops/assets) using the `Executor` defined on the `Job`. If the `RunLauncher` `supports_check_run_worker_health` then the `dagster-daemon` process will occasionally check the health of that launched process but not in a way that is expected to consume continued cpu/mem. 

If instead this language is trying to talk about the code that is in `dagster-celery` that is used by the `celery_executor` then maybe its more correct. The core execution loop manages a `celery.AsyncResult` for each concurrent step, so will scale in some capacity by the number of concurrent steps in a single job but I expect it to be relatively small. I would be surprised if this was problematic at the scale of hundreds or thousands. 
 
I think to continue this discussion it would be most useful to link to specific code of concern or even better profiler results that highlight specific problems. 


https://docs.dagster.io/deployment/executors#executors
https://docs.dagster.io/deployment/run-launcher#run-launchers
